### ISSUE TITLE
turbolinks "instantclick"

### ISSUE THREAD

**Issue by Enalmada** [turbolinks/turbolinks#313]
I want to use turbolinks but there is something it isn't doing that others seem to be doing...fetching content on hover and using that on click.  

This is the main idea behind instantclick.io and the concept seems critical to unlocking the full performance potential of turbolinks.  Note that there seems to be some past discussion about prefetching in general that went to dark places (https://github.com/turbolinks/turbolinks/issues/84) talking about plugins and not universally supported hints.  I feel like hints and plugins are overboard...I feel like doing the same thing as instaclick deserves to be natively supported...even the default behavior (with an opt-out attribute for server side analytics or mutable links).  

If turbolinks did a fetch of link on hover (rather than click), and used that content on click, it would reduce up to several hundred milliseconds of latency for the average user.  It is hard to believe at first but it is true...try it for yourself: http://instantclick.io/click-test.    For a properly tuned backend just dealing with network latency, that amount of time can be the difference in user perception between a site being fast and being instant.   

So turbolinks, can you please consider doing this?

**Comment by joeldrapper** [comment]
I really like this idea! InstantClick feels *very* fast compared with Turbolinks, especially on a slow 3G connection.

In some cases, I think it would make sense to load the page into the cache on hover, but still do a normal request on click, in case the content has changed since hovering. Perhaps there should be an option for this?

It would also need to be specifically limited to `GET` requests, as prefetching `POST`, `PATCH`, and `DELETE` links could have unintended consequences.

@sstephenson what are your thoughts?

**Comment by Enalmada** [comment]
If anyone wants a tease at the potential of this, the gist from glennfu works as a drop in example: https://gist.github.com/Enalmada/6e11191c1ea81a75a4f266e147569096. If you were going to use in production you would probably want to modify to not blow out the normal turbolinks cache after lots of hover (to avoid blowing out your backspace position).  In the production version of this concept, I believe by default only the latest hover should be cached (which could be a quick fix to the cache blowout) and hover cache should be cleared after click but all this could be configurable depending on how static a site is.

**Comment by Ch4s3** [comment]
Would it make sense to roll somethin in like @Enalmada's example behind a flag? Then have users either run `Turbolinks.enable_insta_click()`, or perhaps tags links with `<a href='#' data-tubrolinks-insta-click>`.

**Comment by Ch4s3** [comment]
I decided to try this out and put together [this gist](https://gist.github.com/Ch4s3/49a018d32e93bbf7d52064ad9b314d78) based on the one by Enalmada. I added a check for `data-turbolinks-preload="true"` on the link, and it works pretty well. With the added check you get regular turbolinks for all links, and prefetch on selected links. I still need to look into the cache blowout issue.

**Comment by panoply** [comment]
+1 on this, would love to push @Ch4s3 fix to production but am concerned about cache blowout. Is it an idea to clear the cache after x amount of clicks, e.g: call `Turbolinks.clearCache()`?

**Comment by Ch4s3** [comment]
@panoply I think `Turbolinks. clearCache()` after x clicks is possibly correct, but I'd want other folks to weigh in. Also, my gist is in es6 and not coffeescript as I don't really know coffeescript, so we'd need to caffeinate the code. I'd be happy to work on this, but I'd need some input on coffeescript and the codebase.

**Comment by panoply** [comment]
@Ch4s3 I fall into the same boat, coffeescript is yet to become apart of my knowledge stack. In theory one could easily log x click in conjunction with LocalStorage. It's adding that extra layer but is how I would initially tackle it, going further, one could even prevent the preload running if page already exists in cache (i'm not sure if this is something already included by default). Your current edition increased page-to-page visits by 200ms – 300ms for me.

**Comment by Enalmada** [comment]
I believe that a low risk high reward default could be precaching only the current hover (re-caching that hover if user moves away and hovers over it again).  Average users would feel the extreme "instaclick" performance sensation with negligible risk of stale content.   Sites that have mostly mutable links could easily disable it, static content (ex: blog) sites could increase cache size or allow cache to span multiple pages.  

```
precache-size: 1 (set to 0 to disable feature).  Note: value of 1 fetches fresh content with every hover for balance between latency and risk of stale content.  Increase cache size to 2+ to enable cache to persist beyond hover.)
clear-cache-on-click: true (to ensure min risk of stale content by default.  ie: user hovers a link, posts a comment to site, clicks a link expecting to see their comment)
To disable preload on a per-link basis, add this: data-turbolinks-preload="false"
```

(I mean...the gists floating around ALMOST work...the final version just needs make sure precache wont ever blow out normal cache so users don't lose back button page position if they go hover crazy.  I assume it might be best to implement with a completely separate precache storage instance that is checked first rather than shared.)

Bottom line...could someone with turbolinks experience help take this from gist to something more official?  I don't know turbolinks or coffee but after experiencing how amazing "instaclick" makes a site feel with just around 100 lines of code, this is too important to let languish.

**Comment by Ch4s3** [comment]
@panoply 
> Your current edition increased page-to-page visits by 200ms – 300ms for me.

That sounds about like the latency of the server round trip, so its working as expected 👍 

@Enalmada yeah, I'd be in favor of something low risk, and the more I think about it, only precaching 1 link at a time would be good. 

>  Sites that have mostly mutable links could easily disable it

I'd probably prefer to have this be a per link opt in, like my gist. that makes it super narrow, which I like, but you could add it programmatically too.

> could someone with turbolinks experience help take this from gist to something more official

As stated previously, I'd be happy to help, but I need some direction.

**Comment by Enalmada** [comment]
I guess whether you prefer opt-in or opt-out totally depends on what percentage of your site links have side-effects (or server analytics).  I have no doubt some flags could be created to flip it one way or the other to make everyone happy.  I wonder what is the best for the average user though.  If I had to guess I would say the average link doesn't have side effects...so keep code [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself) and follow the turbolinks pattern.

**Comment by panoply** [comment]
@Enalmada I agree. Let's take a webshop for example, a user browsing the "products" page are going to be hovering over each product image within a grid and without an opt-in / opt-out this would make for a lot of unnecessary requests. In the use-case scenario an `opt-in` would be something one would most definitely require. It may also be an idea to possibly integrate a `setTimeout()` before the `prefetch` initializes, for example if users mouse is hovering a`href` over 1500ms, then begin the prefetch.

Really interested in exploring this further.

**Comment by Enalmada** [comment]
@panoply That totally makes sense and it will probably be common for people to have special sections of the site they want to preload with care or not at all.  The important point here is that hopefully choice can be up to them without making implementation/documentation any more complicated for the majority case.

I am just thinking aloud here but perhaps the right thing to do is not define the touchstart/mouseover event handlers in the plugin but allow people to define their own.
```
$("a:not(.opt-out)").addEventListener("mouseover", preload);
$("a.opt-in").addEventListener("mouseover", preload);
```

Perhaps for the special cases, having an example of something like this could be interesting: [jquery-hoverIntent](https://briancherne.github.io/jquery-hoverIntent/).

**Comment by panoply** [comment]
@Ch4s3 not currently. I'd like to see this feature integrated but I am not prepared to learn or get familiar with coffeescript, especially with the advances in ES7 and ES8. I can confirm the methods I have tried in conjunction with Turbolinks increased performance, even the click to action time ratio was increased with preloading on hover, but adding an extra layer on top of Turbolinks is always a mission in itself, the maintainers are super picky.

**Comment by Ch4s3** [comment]
@panoply Yeah I'm aware of their general selectiveness. I wonder if it makes sense then to release this as a plugin of sorts. Maybe an npm package and or gem that adds this as a layer on top. That way the maintainers here don't have to take it on in the short run, and we could roll it in ES7.

**Comment by Ch4s3** [comment]
@panoply I'll read up on best practices for starting new npm packages and ping you once I have something going.

**Comment by jpic** [comment]
I thought we already had that in Turbolinks :joy:

**Comment by Enalmada** [comment]
FYI Until instantclick behavior is turbolinks native (or plugin) ready, you could consider doing a prefetch of any link a user stops moving over:

```
// https://briancherne.github.io/jquery-hoverIntent/
$(".prefetch").hoverIntent(prefetch);

function prefetch() {
        var link = $(this).attr("href");
        if (typeof link === typeof undefined || link === false) {
            link = $(this).attr("data-href");
        }
        var prerenderLink = $("#prefetchLink");
        if (prerenderLink.length) {
            if (prerenderLink.attr("href") === link) return;
            prerenderLink.attr("href", link);
        } else {
            $('<link id="prefetchLink" rel="prefetch" href="' + link + '" />').appendTo("body");
        }
}
```
(Inspired by [this question](https://stackoverflow.com/questions/32336508/prefetch-prerender-multiple-pages-on-hover))

Note that pages must be browser cacheable to have benefit so dynamic user content pages will need to be something like Cache-Control: private, max-age=5.  

Todo: hoverintent adds some delay so make it optional by only using it if loaded or opt out class (.prefetch-immediate), offer opt in/out by letting people submit selector (ie a:not[.no-prefetch]), ignore invalid link like hash or javascript placeholder, offer rel=prerender opt in for people who understand the consequences.  

Perhaps just making this more robust and into a plugin could be the best thing to do for now for the turbolinks community.

**Comment by grosser** [comment]
Cleanup up version of above:
 - removing the data-href because idk what it does and 
 - docs
 - only prefetch paths and not mailto or outside urls
 - also prefetch after turbolinks transition
 - apply to all turbolinks links

```js
// Using https://briancherne.github.io/jquery-hoverIntent/
// and turbolinks prefetch to make clicks instant
(function(){
  var $prefetcher;

  function prefetch() {
    var href = $(this).attr("href");
    if(!href.match(/^\//)){ return; } // do not prefetch outside urls or mailto:

    // add or change the prefetched link, be careful not to preload the same href multiple times
    if ($prefetcher) {
      if($prefetcher.attr("href") != href) {
        $prefetcher.attr("href", href);
      }
    } else {
      $prefetcher = $('<link rel="prefetch" href="' + href + '" />').appendTo("body");
    }
  }

  // on page-load and after turbolinks transition
  $(document).on("turbolinks:load", function(){
    $("a[data-turbolinks!=false]").hoverIntent(prefetch);
  });
}());
```

took our site from ~140 mouse-down-to-paint to ~40ms 🎉  (rest is mostly paint + js overhead)

![screen shot 2018-06-08 at 9 34 21 am](https://user-images.githubusercontent.com/11367/41170297-833f0b0a-6b00-11e8-8e55-4d35a1c8c5cc.png)
![screen shot 2018-06-08 at 9 35 12 am](https://user-images.githubusercontent.com/11367/41170298-83583a6c-6b00-11e8-829d-20067bf0d8dd.png)

**Comment by danielcompton** [comment]
There is also https://github.com/tristen/hoverintent which is a port of jQuery-hoverintent.

**Comment by Ch4s3** [comment]
I still need to dig into this as a separate project, it's on my TODO list.

**Comment by hopsoft** [comment]
Here's my stab at this. It has no dependencies. I'd love some extra eyeballs to verify that it works properly. https://gist.github.com/hopsoft/ab500a3b584e2878c83137cb539abb32

Some details on the gist. 

- Anchors must explicitly opt-in to prefetch behavior. `<a data-prefetch='true' ...>`
- The hover delay is hard coded at `600ms`

**Comment by jpic** [comment]
@hopsoft thanks, i thought it was working for me, but then started to find this kind of errors 

```
[Learn More] index.js:94:8
../../../src/crudlfap/js/index.js/</fetchers[url]</<
index.js:94:8
[Show/hide message details.] TypeError: window.Turbolinks.Snapshot.fromElement is not a function[Learn More] index.js:94:8
Will-change memory consumption is too high. Budget limit is the document surface area multiplied by 3 (702832 px). Occurrences of will-change over the budget will be ignored. mrsrequest
[Show/hide message details.] TypeError: window.Turbolinks.Snapshot.fromElement is not a function[Learn More] index.js:94:8
```

**Comment by jpic** [comment]
My bad, this method has been renamed to from fromHTMLElement https://github.com/turbolinks/turbolinks/commit/84303b9f92a8cce86c4ebd26554d81efe636961c

**Comment by hopsoft** [comment]
@jpic Thanks. I updated the [gist](https://gist.github.com/hopsoft/ab500a3b584e2878c83137cb539abb32).

**Comment by jpic** [comment]
@hopsoft thank *you* for that :joy:

I got it to work, but [changed it to my needs](https://github.com/yourlabs/crudlfap/blob/master/js/index.js)

For me this still downloads the page when the user clicks, even if it already downloaded it during hover prefetch. I suppose this just means that turbolinks uses cache for preview and then refresh when it gets the page ? That's not blocking but could be avoided maybe ? Or am I getting this wrong ? 

Thanks in advance for sharing some of your insight, I love this feature :joy:

**Comment by hopsoft** [comment]
Ah interesting. Yeah the preview makes it feel faster to the user but we could likely avoid the second request altogether. Need to think through how to best accomplish that.

**Comment by talhashoaib** [comment]
@panoply can you share the @Enalmada version with your edits? I need to use it in production as well.

**Comment by panoply** [comment]
The version we have is tied deeply integrated into our application. What I mean by this is certain actions within the application, reset and/or set preload cache, I need to find the time to break it apart is make it more universal for others to use.

**Comment by daau** [comment]
For the curious, here's an [example of a Rails app](https://dev.to/) that uses instantclick instead of turbolinks. I couldn't believe it was a Rails app with server side rendering when I first used it.

**Comment by hopsoft** [comment]
I've updated the code I've been using for this. Note: This solution works with Safari also. https://gist.github.com/hopsoft/ab500a3b584e2878c83137cb539abb32

**Comment by domchristie** [comment]
Just throwing [GoogleChromeLabs/quicklink](https://github.com/GoogleChromeLabs/quicklink) into the mix. It uses a different approach in that it prefetches all links in the viewport, but has some nice [features](https://github.com/GoogleChromeLabs/quicklink#how-it-works) which might be worth integrating:

> - Waits until the browser is idle (using requestIdleCallback)
> - Checks if the user isn't on a slow connection (using navigator.connection.effectiveType) or has data-saver enabled (using navigator.connection.saveData)

**Comment by grosser** [comment]
Prefetching all links sounds kinda wasteful especially on larger pages and
in my experience preloading on hover already gives instant feedback, so no
advantage in doing more extreme things

**Comment by domchristie** [comment]
> Prefetching all links sounds kinda wasteful especially on larger pages

I agree, and the `hoverintent` solutions above are certainly preferable to quicklink on its own. However, the solutions might be enhanced with the quicklink features listed above. I'm thinking about touch devices in particular. `hoverintent` relies on mouse position, so it has no benefit to touch devices. However, on a small touch screen, where fewer links are visible, quicklink might come in handy.

**Comment by jpic** [comment]
@hopsoft does it still do two requests instead of one ?
I wonder if ServiceWorker is not the best way to prevent the second request. I've done some service worker experiments with a non-turbolinks project and it works extremely well.

**Comment by sfusato** [comment]
The developer of InstantClick also released a newer, lighter plugin called InstantPage: 
https://github.com/instantpage/instant.page

What would be the best way to integrate this with Turbolinks?

**Comment by hopsoft** [comment]
@jpic I've since updated the prefetch script I'm using. https://gist.github.com/hopsoft/ab500a3b584e2878c83137cb539abb32

The prefetcher only makes 1 request. It ensures that clicking a prefetched url is handled as a Turbolinks restoration visit. A 2nd request is still made to pick up any page changes since the prefetch cached the Turbolinks snapshot... as per normal Turbolinks behavior.

Perceived load time is very fast as it is with any Turbolinks restoration visit.

I should reiterate that this solution has no dependencies and also works in Safari.

**Comment by david-a-wheeler** [comment]
@hopsoft .. cool! Can you turn that into a pull request to be included in turbolinks? As long as it can be controlled on a per link basis, I think that would be worth doing.

**Comment by charnould** [comment]
Any update on this pull request?
It would be super nice to have this feature in Turbolinks.
Thanks

**Comment by panoply** [comment]
@hopsoft this is clean. Very clean.

**Comment by frederikhors** [comment]
@hopsoft your work (here: https://github.com/turbolinks/turbolinks/issues/313#issuecomment-482840651) is amazing.

@sstephenson, @packagethief, @domchristie can we PR that code in a useful feature requested by community for years now?

Not everything must always be perfect immediately, I believe, but if we start today then fix/nth PRs will come tomorrow.

Thanks for your work.

**Comment by dieulot** [comment]
Author of [instant.page](https://instant.page) here. Wouldn’t using both Turbolinks and instant.page satisfy your need? I don’t see why they wouldn’t be compatible, and instant.page is 1kB and loaded after everything else, unlike if it would be bundled with Turbolinks.

**Comment by senordelaflor** [comment]
> @jpic I've since updated the prefetch script I'm using. https://gist.github.com/hopsoft/ab500a3b584e2878c83137cb539abb32
> 
> The prefetcher only makes 1 request. It ensures that clicking a prefetched url is handled as a Turbolinks restoration visit. A 2nd request is still made to pick up any page changes since the prefetch cached the Turbolinks snapshot... as per normal Turbolinks behavior.
> 
> Perceived load time is very fast as it is with any Turbolinks restoration visit.
> 
> I should reiterate that this solution has no dependencies and also works in Safari.

@jpic are you using this in production? love it!

**Comment by hopsoft** [comment]
I've got it in production on a few apps. So far so good.

**Comment by frederikhors** [comment]
@hopsoft, `let` can be changed to `const` on lines 6, 22, 48.

`'url' is never reassigned. Use 'const' instead.eslint(prefer-const)`.

**Comment by panoply** [comment]
While this feature is great and I've used a variation of it in a past project bundling it with Turbolinks might not be the best idea and I can understand why maintainers would be hesitant. 

Sometimes features like this become troublesome when being used irresponsibly. Applying it to elements that users are hovering but not clicking like a products page where scrolling, stopping and hovering is abundant leads to countless requests being processed in vain and though this can be prevented with conditional execution using data attributes and timeouts, developers (especially those new the game) tend to care little for the cost of speed increases. 

One thing about projects like Turbolinks and Stimulus I've noticed is how vigorously thought through in both design and functionality.  It would be interesting to see if something like this is merged or extended upon.

**Comment by david-a-wheeler** [comment]
@dieulot - I've been assuming that Turbolinks and instant.click are not compatible, since Turbolinks fundamentally changes click behavior and maintains its own cache.  If they can simply be installed simultaneously that'd be great, but I don't want to assume that.

**Comment by mayrsascha** [comment]
FYI there is a thread on [instant.page](https://instant.page) 's issue tracker regarding compatibility with Turbolinks https://github.com/instantpage/instant.page/issues/52

**Comment by feliperaul** [comment]
@hopsoft Thank you so much for your snippet. But one question: aren't you getting the security warnings on Rails 5.2 ?

I'm getting:

```
Security warning: an embedded <script> tag on another site requested protected JavaScript. If you know what you're doing, go ahead and disable forgery protection on this action to permit cross-origin JavaScript embedding
```

I don't understand why it's triggering this tough, since it's same origin.

**Comment by phacks** [comment]
I wrote a blog post on my experience setting up InstantClick with Rails (and adding a custom cache to avoid repeating requests on multiple hovers of the same link) in case anybody’s interested: https://dev.to/phacks/the-rocky-road-to-implementing-link-prefetching-in-rails-oo0

**Comment by huacnlee** [comment]
I wrote a Npm for implement prefetch for Turbolinks:

> It's reference to this [gist](https://gist.github.com/hopsoft/ab500a3b584e2878c83137cb539abb32)

https://github.com/huacnlee/turbolinks-prefetch

```js
import Turbolinks from 'turbolinks';
window.Turbolinks = Turbolinks;

import TurbolinksPrefetch from 'turbolinks-prefetch';
TurbolinksPrefetch.start();
```

![](https://user-images.githubusercontent.com/5518/94414149-92935b00-01ae-11eb-9916-778d7740db98.png)
